riscyforth.1
.\" Automatically generated by Pandoc 2.17.1.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "RISCYFORTH" "" "August 2023" "" ""
.hy
.PP
RICSCYFORTH(1) RISCYFORTH MANUAL RISCYFORTH(1)
.SH NAME
.PP
riscyforth - Forth for RISC-V SBCs
.SH SYNOPSIS
.PP
riscyforth riscyforth [-h | -u] riscyforth -v riscyforth -i filepath
.SH DESCRIPTION
.PP
\f[B]riscyforth\f[R] is a 64-bit Forth for RISC-V single board
computers.
It is mostly written in RISC-V assembly and is free software with
modification and distribution (including of this man file) covered by
version 2 of the GNU General Public Licence (which should be found in
every distribution of \f[B]riscyforth\f[R]), or any later version, at
your discretion.
.PP
The most up-to-date version of \f[B]riscyforth\f[R] can be found at
https://github.com/mcmenaminadrian/riscyforth
.PP
Staring \f[B]riscyforth\f[R] will invoke a terminal interface and
programs may be loaded or written directly.
Persistent programs need to be written in an editor and loaded into a
running instance of \f[B]riscyforth\f[R].
.PP
This man file describes the words natively supported by
\f[B]riscyforth\f[R].
Like all Forths \f[B]riscyforth\f[R] may be extended through \f[I]colon
words\f[R] written in Forth itself.
.PP
\f[B]riscyforth\f[R] aims to support the Forth 2012 standard but not all
words here are part of the standard and not quite all of the standard is
(currently) supported.
.PP
It should be noted that \f[B]riscyforth\f[R] is a 64 bit Forth - i.e.,
cells are 8 bytes wide by default.
.PP
\f[B]riscyforth\f[R] is copyright (c) Adrian McMenamin, 2020 - 2024.
.SH OPTIONS
.IP
.nf
\f[C]
    -h, -u
        Show usage message

    -v
        Show verbose output (default is verbose output is suppressed).

    -i filepath
        Include Forth file found at filepath
\f[R]
.fi
.SH EXAMPLES
.IP
.nf
\f[C]
    riscyforth -i ./src/exceptions.fth
        Load the Forth exceptions code on startup from the src directory.
\f[R]
.fi
.SH FORTH WORDS
.PP
An alphabetical list of core Forth words implemented on
\f[B]riscyforth\f[R] follows.
A word in the Forth 2012 Core which is not yet implemented (and may
never be) is also noted for the convenience of those porting code.
.PP
\f[B]ABORT\f[R] Quit without displaying a message
.PP
\f[B]ABORT\[rq]\f[R] (x \[en]) if \f[I]x\f[R] is non-zero quit
displaying a message e.g., \f[I]1 ABORT\[rq] Display this message on
quitting\[rq]\f[R]
.PP
\f[B]ABS\f[R] (n \[en] u) Place the absolute value of \f[I]n\f[R] on the
stack
.PP
\f[B]ACCEPT\f[R] (addr n1 \[en] n2) Store at most \f[I]n1\f[R]
characters in buffer at address \f[I]addr\f[R], \f[I]n2\f[R] is the
actual number of characters stored
.PP
\f[B]ACTION-OF\f[R] (\[en] xt) At run time returns the execution token
\f[I]xt\f[R] for the deferred word eg., to use \f[I]ACTION-OF
TESTDEFER\f[R]
.PP
\f[B]AGAIN\f[R] Return to the code block marked by \f[I]BEGIN\f[R]
.PP
\f[B]ALIGN\f[R] (\[en]) If required (8-byte) align the data space
pointer
.PP
\f[B]ALIGNED\f[R] (a1 \[en] a2) \f[I]a2\f[R] is the first aligned
address greater than or equal to \f[I]a1\f[R]
.PP
\f[B]ALLOCATE\f[R] (u \[en] addr ior) allocate \f[I]u\f[R] bytes at
address \f[I]addr\f[R]
.PP
\f[B]ALLOT\f[R] (n \[en]) move the data space pointer by \f[I]n\f[R]
bytes
.PP
\f[B]AND\f[R] (x1 x2 \[en] x3) \f[I]x3\f[R] is the bitwise logical
\f[I]AND\f[R] of \f[I]x1\f[R] with \f[I]x2\f[R]
.PP
\f[B]BASE\f[R] (\[en] a) \f[I]a\f[R] is the address of the current
number conversion radix (number base)
.PP
\f[B]BBLACK\f[R] Set terminal background to black
.PP
\f[B]BBLUE\f[R] Set terminal background to blue
.PP
\f[B]BCYAN\f[R] Set terminal background to cyan
.PP
\f[B]BEGIN\f[R] Begins a loop code block
.PP
\f[B]BGREEN\f[R] Sets terminal background to green
.PP
\f[B]BINARY\f[R] Set the radix (base) to two
.PP
\f[B]BL\f[R] (\[en] c) Places the value 0x20 (ascii 32 - space) on the
stack
.PP
\f[B]BLACK\f[R] Set terminal foreground to black
.PP
\f[B]BLUE\f[R] Set terminal foreground to blue
.PP
\f[B]BMAGENTA\f[R] Set terminal background to magenta
.PP
\f[B]BRED\f[R] Set terminal background to red
.PP
\f[B]BRIGHT\f[R] Set terminal output to bright
.PP
\f[B]BUFFER:\f[R] (x \[en] a) Creates a buffer of length x that returns
address \f[I]a\f[R] e.g., \f[I]80 BUFFER: TESTBUF\f[R] creates the
buffer \f[I]TESTBUF\f[R] of length 80 bytes
.PP
\f[B]BWHITE\f[R] Set terminal background to white
.PP
\f[B]BYE\f[R] Leave \f[B]riscyforth\f[R]
.PP
\f[B]BYELLOW\f[R] Set terminal background to yellow
.PP
\f[B][\f[R] At compile time the contents of \f[I][ ]\f[R] are
immediately evaluated
.PP
\f[B][CHAR]\f[R] (\[lq]spaces<name>\[rq] \[en]) At runtime: (\[en] char)
Places value of first character in \f[I]name\f[R] on stack as
\f[I]char\f[R]
.PP
\f[B][COMPILE]\f[R] NOT IMPLEMENTED: Word is obsolescent in Forth 2012.
Use \f[I]POSTPONE\f[R] instead.
.PP
\f[B][\[cq]]\f[R] (\[lq]spaces\[rq]<name>\[rq] \[en]) At runtime: (\[en]
xt) Please use \f[I]\[cq]\f[R] instead (\f[I][\[cq]]\f[R] has the same
functionality)
.PP
\f[B]CASE\f[R] Mark the start of \f[I]CASE ..
OF ..
ENDOF ..
ENDCASE\f[R] control structure
.PP
\f[B]C,\f[R] (c \[en]) Advance data space pointer by 1 and store
\f[I]c\f[R] in data space
.PP
\f[B]CELL+\f[R] (a1 \[en] a2) add a cell size (8 bytes) to \f[I]a1\f[R]
and store sum \f[I]a2\f[R] on the stack
.PP
\f[B]CELLS\f[R] (n1 \[en] n2) store the size, \f[I]n2\f[R] of
\f[I]n1\f[R] cells on the stack
.PP
\f[B]C\[at]\f[R] (c-addr \[en] char) Fetch the character stored at
\f[I]c-addr\f[R] and store on the stack as \f[I]char\f[R]
.PP
\f[B]CHAR\f[R] (\[lq]name\[rq] \[en] char) Put the first character of
the string \f[I]name\f[R] on the stack as \f[I]char\f[R]
.PP
\f[B]CHAR+\f[R] (caddr1 \[en] caddr2) Add character size (1) to address
\f[I]caddr1\f[R] and store result on stack in \f[I]caddr2\f[R]
.PP
\f[B]CHARS\f[R] (n1 \[en] n2) \f[I]n2\f[R] is the size in address units
of \f[I]n1\f[R] (this is a NOP)
.PP
\f[B]COMPILE,\f[R] (xt \[en]) At compile time \f[I]xt\f[R] is compiled
in (replacing \f[I]COMPILE,\f[R])
.PP
\f[B]CONSTANT\f[R] Create a word that returns a constant value to the
stack e.g., \f[I]25 FIVESQUARED CONSTANT\f[R] creates the constant
\f[I]FIVESQUARED\f[R] that will always return 25 on the stack
.PP
\f[B]COUNT\f[R] (c-addr1 \[en] c-addr2 u) Return character count and
text address for counted string
.PP
\f[B]CR\f[R] (\[en]) output a newline
.PP
\f[B]CREATE\f[R] Create a word that returns a constant pointer to the
data space e.g., \f[I]CREATE TEST\f[R] creates the word \f[I]TEST\f[R]
that returns the value of the data space pointer at the time of
creation.
.PP
\f[B]C!\f[R] (char caddr \[en]) Stores character \f[I]char\f[R] at
\f[I]caddr\f[R]
.PP
\f[B]CYAN\f[R] Set termainal foreground to cyan
.PP
\f[B]:\f[R] Begin a \f[I]colon word\f[R] definition
.PP
\f[B]:NONAME\f[R] (\[en] xt) Create a \f[I]colon-word\f[R] and place
execution token \f[I]xt\f[R] on stack
.PP
\f[B],\f[R] (x \[en]) Advance data space pointer by one cell and store
\f[I]x\f[R] in the cell
.PP
\f[B]C\[rq]\f[R] (\[en] c-addr) on execution, (\[lq]ccc\[rq] \[en]) on
compilation.
Return counted string at address \f[I]c-addr\f[R] - compiled code only.
Does nothing in intrepreted code.
.PP
\f[B]CRESET\f[R] (mask addr \[en]) Turn bits off at \f[I]addr\f[R] using
8-bit \f[I]mask\f[R]
.PP
\f[B]CROSSEDOUT\f[R] ( \[em] ) Cross out text
.PP
\f[B]CSET\f[R] (mask addr \[en]) Set bits at \f[I]addr\f[R] using 8-bit
\f[I]mask\f[R]
.PP
\f[B]CTOGGLE\f[R] (mask addr \[en]) Toggle bits at \f[I]addr\f[R] using
8-bit \f[I]mask\f[R]
.PP
\f[B]CUBE\f[R] (x1 \[en] x2) Cube \f[I]x1\f[R] and store in \f[I]x2\f[R]
.PP
\f[B]DECIMAL\f[R] Set radix (base) to ten
.PP
\f[B]DEFER\f[R] Defer execution of created word to another word e.g.,
\f[I]DEFER TEST\f[R] creates a word \f[I]TEST\f[R] that we can later
assign execution characteristics to (see e.g., \f[I]DEFER\[at]\f[R])
.PP
\f[B]DEFER\[at]\f[R] (x1 \[en] x2) Reports that execution token
\f[I]x1\f[R] is set to \f[I]x2\f[R] e.g., \f[I]EMIT IS TESTDEFER \[cq]
TESTDEFER DEFER\[at] \[cq] EMIT =\f[R] will return \f[I]TRUE\f[R] if a
call to \f[I]TESTDEFER\f[R] executes \f[I]EMIT\f[R]
.PP
\f[B]DEFER!\f[R] (x2 x1 \[en]) will set execution token \f[I]x1\f[R] to
\f[I]x2\f[R] e.g., \f[I]\[cq] EMIT \[cq] TESTDEFER DEFER!\f[R] will set
\f[I]TESTDEFER\f[R] to execute \f[I]EMIT\f[R]
.PP
\f[B]DEPTH\f[R] (\[en] n) Reports depth of stack
.PP
\f[B]DISPLAY\f[R] (x..x \[en] x..x) will display zero terminated string
built from stack
.PP
\f[B]DO\f[R] Begins \f[I]LOOP\f[R] block in form: \f[I]limit first DO
\&...
LOOP\f[R] (or \f[I]+LOOP\f[R] or \f[I]-LOOP\f[R])
.PP
\f[B]DOES>\f[R] Assigns execution body to word created in data space
e.g., \f[I]: INDEXED-ARRAY CREATE CELLS ALLOT DOES> SWAP CELLS + ;\f[R]
creates an indexed array type
.PP
\f[B]DROP\f[R] (x \[en])
.PP
\f[B]DUP\f[R] (x \[en] x x)
.PP
\f[B]/\f[R] (n1 n2 \[en] n3) Divide \f[I]n1\f[R] by \f[I]n2\f[R] and
store the result on the stack as \f[I]n3\f[R]
.PP
\f[B]/MOD\f[R] (n1 n2 \[en] n3 n4) Divide \f[I]n1\f[R] by \f[I]n2\f[R],
storing the remainder \f[I]n3\f[R] and the quotient \f[I]n4\f[R] on the
stack
.PP
\f[B].R\f[R] (n1 n2 \[en]) Display \f[I]n1\f[R] right flushed in a field
of width \f[I]n2\f[R]
.PP
\f[B].S\f[R] Debug word that displays contents of stack
.PP
\f[B].(\f[R] (\[lq]ccc\[rq] \[en]) Parse and display ccc - immediate
word
.PP
\f[B].\[rq]\f[R] Output the enclosed string e.g.\ \f[I].\[rq] Output
this\[rq]\f[R]
.PP
\f[B]DROPINPUT\f[R] Discard rest of input line
.PP
\f[B]ELSE\f[R] \f[I]ELSE\f[R] clause in \f[I]IF ..
ELSE ..
THEN\f[R]
.PP
\f[B]EMIT\f[R] (x \[en]) Output character of value \f[I]x\f[R]
.PP
\f[B]ENDCASE\f[R] Mark the end of \f[I]CASE ..
OF ..
ENDOF ..
ENDCASE\f[R] control structure
.PP
\f[B]ENDOF\f[R] Mark the end of \f[I]OF ..
ENDOF\f[R] clause in \f[I]CASE ..
ENDCASE\f[R] control structure
.PP
\f[B]ENVIRONMENT?\f[R] (addr u \[en] false | i * x true) Query the local
environment.
\f[I]addr\f[R] and \f[I]u\f[R] are the address and length of a query
string.
Returns false if query is not supported, otherwise an answer based on
the query.
Currently supported query words: /COUNTED-STRING /HOLD /PAD
ADDRESS-UNIT-BITS FLOORED MAX-CHAR MAX-D MAX-N MAX-U MAX-UD
RETURN-STACK-CELLS STACK-CELLS
.PP
\f[B]ERASE\f[R] (addr u \[en]) if \f[I]u\f[R] greater than zero, clear
(set to zero) \f[I]u\f[R] bytes from address \f[I]addr\f[R]
.PP
\f[B]EVALUATE\f[R] ( i * x c-addr u \[en] j * x ) Interpret the string
at \f[I]c-addr\f[R]
.PP
\f[B]EXECUTE\f[R] (xt \[en] ?)
Remove \f[I]xt\f[R] from stack and execute it
.PP
\f[B]EXIT\f[R] Leave an \f[I]IF ..
ELSE ..
THEN\f[R] structure (care must be taken to \f[I]UNLOOP\f[R] if
necessary)
.PP
\f[B]=\f[R] (x1 x2 \[en] flag) Set \f[I]flag\f[R] to \f[I]TRUE\f[R] (-1)
if \f[I]x1 = x2\f[R] otherwise set \f[I]flag\f[R] to \f[I]FALSE\f[R]
(zero)
.PP
\f[B]FALSE\f[R] (\[en] 0) Zero indicating logical false
.PP
\f[B]FILL\f[R] (c-addr u char \[en]) if \f[I]u\f[R] is greater than zero
set \f[I]u\f[R] bytes from \f[I]c-addr\f[R] onwards to \f[I]char\f[R]
.PP
\f[B]FIND\f[R] (c-addr \[en] caddr 0 | xt 1 | xt -1) Find the definition
named in the counted string at c-addr.
If the definition is not found, return c-addr and zero.
If the definition is found, return its execution token xt.
If the definition is immediate, also return one (1), otherwise also
return minus-one (-1).
.PP
\f[B]FM/MOD\f[R] (d n2 \[en] n1 n0) Floored division: divide \f[I]d\f[R]
by \f[I]n2\f[R] and report remainder in \f[I]n1\f[R] and floored
quotient in \f[I]n0\f[R]
.PP
**\[at]** (addr \[en] x) Fetch as \f[I]x\f[R] the contents of cell at
\f[I]addr\f[R] and store on the stack
.PP
\f[B]FREE\f[R] (addr \[en] ior) Free memory at \f[I]addr\f[R]
(\f[I]ior\f[R] is zero on success)
.PP
\f[B]GETLINE\f[R] Fetch text input
.PP
\f[B]GREEN\f[R] Set terminal foreground to green
.PP
\f[B]HERE\f[R] (\[en] addr) Return the current value of the data space
pointer
.PP
\f[B]HEX\f[R] Set the radix (base) to sixteen
.PP
\f[B]HOLD\f[R] (char \[en]) Add \f[I]char\f[R] to the beginning of a
pictured numeric output string
.PP
\f[B]HOLDS\f[R] (caddr u \[en]) Add counted string defined by
\f[I]u\f[R] and \f[I]caddr\f[R] to the start of pictured numeric output
.PP
\f[B]I\f[R] (\[en] n) Place the value of the current innermost loop
counter on the stack
.PP
\f[B]IF\f[R] (x \[en]) Begin \f[I]IF \&...
ELSE ..
THEN\f[R] structure.
\f[I]IF\f[R] clause is executed if \f[I]x\f[R] is non-zero, otherwise
\f[I]ELSE\f[R] clause (if present) is executed.
.PP
\f[B]IMMEDIATE\f[R] (\[en]) Make the most recent definition an immediate
word
.PP
\f[B]INCLUDE\f[R] Load file and immediately parse e.g.
\f[I]INCLUDE /home/foo/bar.fth\f[R] will load and evaluate
\f[I]bar.fth\f[R]
.PP
\f[B]INVERT\f[R] (x1 \[en] x2) Invert all bits of \f[I]x1\f[R] and store
on the stack as \f[I]x2\f[R]
.PP
\f[B]IS\f[R] (xt \[en]) Set \f[I]name\f[R] to execute \f[I]xt\f[R] e.g.,
\f[I]\[cq] .
IS TESTDEFER\f[R] will ensure \f[I]TESTDEFER\f[R] executes \f[I].\f[R]
.PP
\f[B]J\f[R] (\[en] n) Place the value of the next-outer loop counter on
the stack
.PP
\f[B]KEY\f[R] (\[en] char) Receive one character from the (ASCII)
chaarcter set.
.PP
\f[B]LEAVE\f[R] R:(n \[en]) Immediately leave a loop discarding control
parameters
.PP
\f[B]LITERAL\f[R] (\[en] x) Place \f[I]x\f[R] on the stack e.g.\ \f[I][
x ] LITERAL\f[R]
.PP
\f[B]LOADMODULE\f[R] (\[lq]<spaces>name\[rq] \[en] x) Load the module
\f[I]name\f[R], leave \f[I]TRUE\f[R] on stack on success
.PP
\f[B]LOOP\f[R] Evaluate loop parameters and either terminate loop R:(x
\[en]) or continue to execute loop R:(x1 \[en] x2)
.PP
\f[B]LSHIFT\f[R] (x1 u \[en] x2) left shift \f[I]x1\f[R] by \f[I]u\f[R]
and store as \f[I]x2\f[R]
.PP
\f[B]MAGENTA\f[R] Set terminal foreground to magenta
.PP
\f[B]MARKER\f[R] (\[lq]name\[rq] \[en]) Delete everything in the
dictionary up to and including name
.PP
\f[B]MAX\f[R] (n1 n2 \[en] n3) \f[I]n3\f[R] is the greater of
\f[I]n1\f[R] and \f[I]n2\f[R]
.PP
\f[B]MAXINT\f[R] (\[en] n) Puts 0x7FFFFFFFFFFFFFFF on the stack
.PP
\f[B]MIN\f[R] (n1 n2 \[en] n3) \f[I]n3\f[R] is the lesser of
\f[I]n1\f[R] and \f[I]n2\f[R]
.PP
\f[B]MININT\f[R] (\[en] n) Puts 0x8000000000000000 on the stack
.PP
\f[B]-LOOP\f[R] (n \[en]) R:(x \[en] x1) Subtract \f[I]n\f[R] from
\f[I]x\f[R] and check loop limits
.PP
\f[B]-ROT\f[R] (x2 x1 x0 \[en] x0 x2 x1)
.PP
\f[B]MOD\f[R] (n1 n2 \[en] n3) \f[I]n3\f[R] is the remainder of dividing
\f[I]n1\f[R] by \f[I]n2\f[R]
.PP
\f[B]MOVE\f[R] (addr1 addr2 u \[en]) if \f[I]u\f[R] is greater than zero
copy \f[I]u\f[R] bytes from \f[I]addr1\f[R] to \f[I]addr2\f[R]
.PP
\f[B]M*\f[R] (n n \[en] d) Signed multiplication producing double (128
bit) result
.PP
\f[B]-\f[R] (n1 n2 \[en] n3) \f[I]n3\f[R] is the result of \f[I]n1\f[R]
minus \f[I]n2\f[R]
.PP
\f[B]MS\f[R] (x \[en]) Pause execution for \f[I]x\f[R] milliseconds
.PP
\f[B]NEGATE\f[R] (n1 \[en] n2) \f[I]n2\f[R] is the arithmetic inverse of
\f[I]n1\f[R]
.PP
\f[B]NIP\f[R] (x1 x2 \[en] x2)
.PP
\f[B]OCTAL\f[R] Set radix (base) to eight
.PP
\f[B]OF\f[R] From \f[I]CASE \&...
OF \&...
ENDOF ..
ENDCASE\f[R] structure
.PP
\f[B]OR\f[R] (x1 x2 \[en] x3) \f[I]x3\f[R] is bitwise inclusive or of
\f[I]x1\f[R] with \f[I]x2\f[R]
.PP
\f[B]OVER\f[R] (x1 x2 \[en] x1 x2 x1)
.PP
\f[B]1-\f[R] (x1 \[en] x2) Subtract 1 from \f[I]x1\f[R] and store the
result in \f[I]x2\f[R]
.PP
\f[B]1+\f[R] (x1 \[en] x2) Add 1 to \f[I]x1\f[R] and store the result in
\f[I]x2\f[R]
.PP
\f[B]PAD\f[R] (\[en] addr) return the address of a transient scratch pad
.PP
\f[B]PAGE\f[R] Clear the terminal and set output to the top left
.PP
\f[B]PARSE-NAME\f[R] ( \[lq]name\[rq] \[en] c-addr u ) Skip leading
space delimiters.
Parse name delimited by a space.
.PP
\f[B]PARSE\f[R] ( char \[lq]ccc\[rq] \[en] c-addr u ) Parse ccc
delimited by the delimiter .char
.PP
\f[B]PICK\f[R] (xu \&...
x1 x0 u \[en] x1 x0 xu)
.PP
\f[B]POSTPONE\f[R] (\[lq]<spaces>name\[rq] \[en]) Compile in
\f[I]name\f[R] even if \f[I]IMMEDIATE\f[R] - in general append the
compilation semantics of \f[I]name\f[R] to the current colon definition.
.PP
\f[B]+\f[R] (n1 n2 \[en] n3) \f[I]n3\f[R] is the sum of \f[I]n1\f[R] and
\f[I]n2\f[R]
.PP
\f[B]+LOOP\f[R] (n \[en]) R:(x \[en] x1) Add \f[I]n\f[R] to \f[I]x\f[R]
and check loop limits
.PP
\f[B]+!\f[R] (n addr \[en]) add \f[I]n\f[R] to the value stored in the
cell at \f[I]addr\f[R]
.PP
\f[B]?\f[R] (addr \[en]) Output value stored at address \f[I]addr\f[R]
.PP
\f[B]QUIT\f[R] Leave the executing program
.PP
\f[B]RDROP\f[R] R:(x \[en])
.PP
\f[B]RECURSE\f[R] Re-execute the current word
.PP
\f[B]RED\f[R] Set terminal forground to red
.PP
\f[B]REFILL\f[R] (\[en] flag) Currently just places 0xFFFFFFFFFFFFFFFF
(TRUE) on stack
.PP
\f[B]REPEAT\f[R] End of a \f[I]BEGIN ..
WHILE ..
REPEAT\f[R] block
.PP
\f[B]RESET\f[R] Reset the terminal colours
.PP
\f[B]RESIZE\f[R] (addr1 u \[en] addr2 ior) Copy data at \f[I]addr1\f[R]
to new area of size \f[I]u\f[R] which will be found at \f[I]addr2\f[R]
.PP
\f[B]RESTORE-INPUT\f[R] (xn ..
x1 n \[en] flag) Currently shortens stack by \f[I]n\f[R] entries and
posts TRUE (0xFFFFFFFFFFFFFFFF) only
.PP
\f[B]R\[at]\f[R] (\[en]x) R:(x \[en] x) Copy x from the return stack to
the (data) stack
.PP
\f[B]ROLL\f[R] (xu xu-1 ..
x1 x0 u \[en] xu-1 ..
x1 x0 xu)
.PP
\f[B]ROT\f[R] (x1 x2 x3 \[en] x2 x3 x1)
.PP
\f[B]RSHIFT\f[R] (x1 u \[en] x2) Logically right shift \f[I]x1\f[R] by
\f[I]u\f[R] and store in \f[I]x2\f[R]
.PP
\f[B]R>\f[R] (\[en]x) R:(x\[en]) Move \f[I]x\f[R] from the return stack
to the (data) stack
.PP
\f[B]SAVE-INPUT\f[R] (\[en] 0) Added for completness: merely puts a zero
on the stack.
.PP
\f[B]SIGN\f[R] (n \[en]) if \f[I]n\f[R] is negative add a minus sign to
the beginning of the pictured numeric output string
.PP
\f[B]SM/REM\f[R] (d n2 \[en] n1 n0) Symmetric division: divide
\f[I]d\f[R] by \f[I]n2\f[R] and report remainder in \f[I]n1\f[R] and
symmetric quotient \f[I]n0\f[R]
.PP
\f[B]SOURCE-ID\f[R] (\[en] 0|-1) Places 0 on stack for user-input device
source and -1 for string (via \f[I]EVALUATE\f[R])
.PP
\f[B]SOURCE\f[R] (\[en] addr u) \f[I]addr\f[R] is the address of the
input buffer and \f[I]u\f[R] the number of caharcters it contains
.PP
\f[B]SPACE\f[R] Display one space
.PP
\f[B]SPACES\f[R] (x \[en]) Display \f[I]x\f[R] spaces
.PP
\f[B]SQUARE\f[R] (x1 \[en] x2) Square \f[I]x1\f[R] and store in
\f[I]x2\f[R]
.PP
\f[B]STATE\f[R] (\[en] addr)
\f[I]a\f[R]\f[I]d\f[R]\f[I]d\f[R]\f[I]r\f[R] is the address of a cell
reporting the compilation state
.PP
\f[B]SWAP\f[R] (x1 x2 \[en] x2 x1)
.PP
\f[B];\f[R] Mark the end of a \f[I]colon word\f[R]
.PP
**S\[rs]\[lq]** At compilation: (\[rq]ccc\[rq] \[em]) Parse string ccc
At runtime: (\[en] c-addr u) Address and length of parsed string
.PP
\f[B]S\[rq]\f[R] (\[en] addr u) \f[I]addr\f[R] contains the address of,
and \f[I]u\f[R] the length of, the string defined in the inverted commas
.PP
\f[B]S>D\f[R] (n \[en] d) Convert \f[I]n\f[R] to double \f[I]d\f[R]
.PP
\f[B]!\f[R] (x addr \[en]) Store \f[I]x\f[R] at \f[I]addr\f[R]
.PP
\f[B]TERMIOSSTRING\f[R] (ccc\[rq]string\[rq] \[em] ?)
Interperet string as the body of a termnal escape sequence
.PP
\f[B]THEN\f[R] Final clause in all \f[I]IF\&...\f[R] structures,
execution continues after \f[I]THEN\f[R] once \f[I]IF\f[R] and
\f[I]ELSE\f[R] clauses exhausted
.PP
\f[B]TIB\f[R] (\[en] addr) Returns address of input buffer
.PP

riscyforth.S
#####################################################################################
#           RISCYFORTH                                                              #
#           TIL for RISC-V                                                          #
#           Copyright (c) Adrian McMenamin <adrianmcmenamin@gmail.com> 2020 - 2023  #
#           Licensed for resuse under version 2 of the GPL or any later version     #
#                                                                                   #
#           As per the terms of the GPL no warranty is offered for this code        #
#####################################################################################

.include "./riscylib/riscyvar.S"

   .bss
   .comm SOURCEID, ADDRWIDTH
   .comm NEWWORDENTRY, ADDRWIDTH
   .comm OKSTATUS, ADDRWIDTH
   .comm CALLSTACK, BIGSPACE
   .comm STACK, BIGSPACE
   .comm VARIABLES, BIGSPACE
   .comm CONSTANTS, BIGSPACE
   .comm TOP_VARS, ADDRWIDTH
   .comm BOTTOM_VARS, ADDRWIDTH
   .comm TOP_CONS, ADDRWIDTH
   .comm BOTTOM_CONS, ADDRWIDTH
   .comm inputfileptr, ADDRWIDTH
   .comm DECIDESTACKP, ADDRWIDTH
   .comm DECIDESTACKB, ADDRWIDTH
   .comm VARIABLEFLAG, ADDRWIDTH
   .comm datamanagement, ADDRWIDTH
   .comm recycleallot, ADDRWIDTH
   .comm recyclewatermark, ADDRWIDTH
   .comm dataspace, ADDRWIDTH
   .comm INDOES, ADDRWIDTH
   .comm INPPONE, ADDRWIDTH
   .comm PICTUREDN, 0x10
   .comm INEVAL, ADDRWIDTH
   .comm RESUME_AFTER_EVAL, ADDRWIDTH
   .comm termios, 64
   .comm CYCLESTART, ADDRWIDTH
   .comm CYCLECOUNT, ADDRWIDTH
   .comm CYCLEPINGS, ADDRWIDTH
   .comm INPUT_START_RESTART, ADDRWIDTH
   .comm INPUT_DISPLACE_RESTART, ADDRWIDTH

   .data
   welcome_msg:   	.asciz "RISCYFORTH - Copyright Adrian McMenamin 2020 - 2023"
   riscy_forth0:  	.asciz "*******************************************************************************"
   riscy_forth1:  	.asciz "RISCYFORTH is a FORTH for RISC-V devices."
   riscy_forth3:	.asciz "==============================================================================="
   riscy_forth4:  	.asciz "RISCYFORTH is licenced under version 2 of the GNU General Public Licence or"
   riscy_forth4a:	.asciz "any later version at your discretion."
   riscy_forth5:  	.asciz "See https://github.com/mcmenaminadrian/riscyforth.git"
   riscy_forthQ:  	.asciz "Exiting RISCYFORTH."
   Ok_msg:        	.asciz "OK"
   NotOk_msg:     	.ascii "Input fails with token: "
   .equ NOK_LEN,  24
   bad_colon:	  	.asciz "COLON command failed: please use a key word for your new command."
   zerolen_varin: 	.asciz "Could not read variable (or command) - zero length?"
   failed_varin:  	.asciz "Failed to read variable"
   Error_msg:     	.asciz "INTERPRETER FAILURE! System may be unstable."
   String_err:    	.asciz "ERROR: bad string."
   stdinopen:     	.asciz "r"
   stdoutopen:    	.asciz "w"
   decimalout:    	.asciz "%lld "
   hexout:        	.asciz "%#llX "
   octalout:      	.asciz "%lloo "
   badif:		.asciz "Bad IF"
   badelse:		.asciz "Bad ELSE"
   #ENVIRONMENT? strings
   eq_countedstrings:	.ascii "/COUNTED-STRING"
   .set eq_csLen, .-eq_countedstrings
   eq_hold:		.ascii "/HOLD"
   .set eq_holdLen, .-eq_hold
   eq_pad:		.ascii "/PAD"
   .set eq_padLen, .-eq_pad
   eq_addressunitbits:	.ascii "ADDRESS-UNIT-BITS"
   .set eq_aubLen, .-eq_addressunitbits
   eq_floored:		.ascii "FLOORED"
   .set eq_flooredLen, .-eq_floored
   eq_maxchar:		.ascii "MAX-CHAR"
   .set eq_maxcharLen, .-eq_maxchar
   eq_maxd:		.ascii "MAX-D"
   .set eq_maxdLen, .-eq_maxd
   eq_maxn:		.ascii "MAX-N"
   .set eq_maxnLen, .-eq_maxn
   eq_maxu:		.ascii "MAX-U"
   .set eq_maxuLen, .-eq_maxu
   eq_maxud:		.ascii "MAX-UD"
   .set eq_maxudLen, .-eq_maxud
   eq_returnstackcells:	.ascii "RETURN-STACK_CELLS"
   .set eq_rscLen, .-eq_returnstackcells
   eq_stackcells:	.ascii "STACK-CELLS"
   .set eq_scLen, .-eq_stackcells


  .text
  .align 3
  .globl main
  _start:
  main:
		# handle initial options
		ld a0, 0(sp)					# argc
		ld a1, 8(sp)					# points to first argv element
		call initriscyforth
                la s10, STACK + BIGSPACE - STACKOFFSET
                la s9, CALLSTACK + BIGSPACE - STACKOFFSET
                j launchRISCYFORTH


    .include "macros.S"
    .balign 8
    .include "core.S"
    .balign 8
    .include "basics.S"
    .balign 8
    .include "outerloop.S"
    .balign 8
    .include "transitions.S"
    .balign 8
    .include "search_create.S"
    .balign 8
    .include "decision.S"
    .balign 8
    .include "load.S"
    .balign 8
    .include "coredefs.S"
    .balign 8
    .include "constant.S"
    .balign 8
    .include "memoryallocator.S"
    .balign 8
    .include "memory.S"
    .balign 8
    .include "utilities.S"
    .balign 8
    .include "termioscode.S"
